datreant/backends/core.py

```diff
@@ -316,3 +329,139 @@ def _close(self):
         self._unlock(self.fd)
         self.fdlock = None
         self._close_fd()
+
+
+class FileSerial(File):
+    """File object base class for serialization formats, such as JSON.
+
+    """
+    def _open_file_r(self):
+        return open(self.filename, 'r')
+
+    def _open_file_w(self):
+        return open(self.filename, 'w')
+
+    def read_file(self):
+        """Return deserialized representation of file.
+
+        """
+        self._apply_shared_lock()
+
+        self.handle = self._open_file_r()
+        out = self._deserialize(self.handle)
+        self.handle.close()
+
+        self._release_lock()
+
+    @staticmethod
+    def _read(func):
+        """Decorator for applying shared lock on file.
+
+        Applying this decorator to a method will ensure a shared lock is
+        obtained before that method is executed. It also ensures that the lock
+        is removed after the method returns.
+
+        """
+        @wraps(func)
+        def inner(self, *args, **kwargs):
+            if self.fdlock:
+                out = func(self, *args, **kwargs)
+            else:
+                self._apply_shared_lock()
+                try:
+                    out = func(self, *args, **kwargs)
+                finally:
+                    self._release_lock()
+            return out
+
+        return inner
+
+    @staticmethod
+    def _write(func):
+        """Decorator for applying an exclusive lock on file.
+
+        Applying this decorator to a method will ensure an exclusive lock is
+        obtained before that method is executed. It also ensures that the lock
+        is removed after the method returns.
+
+        """
+        @wraps(func)
+        def inner(self, *args, **kwargs):
+            if self.fdlock == 'exclusive':
+                out = func(self, *args, **kwargs)
+            else:
+                self._apply_exclusive_lock()
+                try:
+                    out = func(self, *args, **kwargs)
+                finally:
+                    self._release_lock()
+            return out
+
+        return inner
+
+    @staticmethod
+    def _pull_push(func):
+        """Decorator for deserializing the contents of file, then reserializing.
+
+        Applying this decorator to a method will ensure the latest version of
+        the data is obtained before that method is executed. It also ensures
+        that changes to the data are written to the file after the method
+        returns.
+
+        This decorator doesn't do any file locking, but should be used after
```

Comment:
Ok this means that the order of decorators might be important. So for example only one of the following two functions uses the right declaration. I'm actually not sure which one it is.

```python
@_read_and_lock
@_pull_push
def foo():
    pass


@_pull_push
@_read_and_lock
def bar():
     pass
```

So I get this is a hard thing to avoid the way you do this right now. But I see that this might be an easy error to make then class methods are declared. A easy thing to do is change the name to something like `_pull_push_needs_lock`, this should work as a reminder for people who review the code that the decorator ordering is important.

Unfortunately I don't know of a way to automatically check the order of applied decorators.

Of course ideally the order doesn't matter but we can't do that here.